### DISCUSSION TITLE
iPXE Framebuffer and Text Console

### DISCUSSION THREAD

**Discussion by rspotoaz** [ipxe/ipxe#207] [Q&A] [answered]
Hello -

I have iPXE booting a script that sets a graphical console w/ a background image.  This works fine unless I'm booting over a serial console (using BIOS console redirection).  As soon as it gets a DHCP IP it just stops.

I have the script set to execute a simple "console" command if the setting of a background image using FB fails (&&), however this is not the solution I need.  In particular, the system I'm working with _has_ a working display chip but access is via serial.  Therefore, the iPXE menu is drawn but not seen.

Is there any way to determine if a serial connection is "active"/"in use" in order to "revert" to a textual console?  I know this is asking a lot.

Thanks for your input!

**Comment by NiKiZe** [comment, marked as answer]
Bios console redirection will require you to use plain text, graphic mode is well, graphic so you won't have text.

Since this is done by bios there is probably no way to detect it.
You could enable iPXEs serial, but then you have to be sure to not have both redirections enabled at the same time.

**Reply by mcb30** [reply]
To elaborate on what @NiKiZe wrote:

iPXE **is** capable of using both graphical and textual consoles simultaneously.  For example, if you have iPXE configured to use all three of `CONSOLE_FRAMEBUFFER`, `CONSOLE_SERIAL` and `CONSOLE_SYSLOG`, then you would see graphical output on the local console, with the same text appearing on the serial port, and with the non-interactive output also sent to a syslog server.

However, in your case you are using BIOS console redirection rather than iPXE's own serial console support.  The BIOS serial console redirection feature (unlike iPXE's serial console support) cannot support this same ability to use graphical and textual consoles simultaneously.

@NiKiZe is correct: you will get the behaviour that you want by disabling the BIOS serial console redirection and enabling iPXE's own serial console support.  The downside is that you will then be able to see only the iPXE output via the serial port; you would lose access to e.g. the BIOS setup screen.

**Reply by rspotoaz** [reply]
Thank you for the reply.  I had assumed as such, but was hoping there was something magical there I could use.

**Reply by mcb30** [reply]
Sorry.  BIOS serial console redirection is hidden behind the `INT 10` API, and there's nothing that provides visibility to iPXE (or other software) as to whether or not it's in use.

If you know that certain systems in your environment will always be configured this way then you may be able to use SMBIOS information to perform platform-specific actions.  For example:

```
goto quirks_${manufacturer} || goto quirks_none
:quirks_Dell
:quirks_HP
console
:quirks_none
```

would reset the console to text mode on Dell or HP systems.

**Reply by rspotoaz** [reply]
Yeah that was my first inclination and looks like the way I'll need to go, at least for my immediate needs.  Thanks again.
